`lightning-liquidity`: Refactor LSPS1 service-side

tnull opened the pull request:
Closes #3480.

We 'refactor' (rewrite) the `LSPS1ServiceHandler`, move state handling to a dedicated `PeerState`, add an STM pattern, add persistence for the service state, add some more critical API paths, add test coverage, and finally remove the `cfg(lsps1_service)` flag.

martinsaposnic committed:
LSPS1: Add initial integration test

We add the first LSPS1 integration test. This is based on the unfinished
work in https://github.com/lightningdevkit/rust-lightning/pull/3864, but
rebased to account for the new ways we now do integration test setup.

tnull committed:
Cleanup unused code

.. for which we got warnings

tnull committed:
Move `PeerState` and related types to `peer_state.rs` module

We move the `PeerState` related types to a new module. In the following
commits we'll bit-by-bit drop the `pub(super)`s introduced here,
asserting better separation of state and logic going forward.

tnull committed:
Replace `insert_outbound_channel` with `PeerState::new_order`

.. requiring less access to internals

tnull committed:
Drop bogus channel state handling

.. we will re-add a proper state machine in a later commit, but for now
we can just drop all of this half-baked logic that doesn't actually do
anything.

tnull committed:
Use `PeerState::{get_order, has_active_orders}` instead of map

Previously, we'd directly access the internal `outbound_` map of
`PeerState`. Here we refactor the code to avoid this.

Note this also highlighted a bug in that we currently don't actually
update/persist the order state in `update_order_state`. We don't fix
this here, but just improve isolation for now, as all state update
behavior will be reworked later.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Use `PeerState::{register,remove}_request` instead of map access

We introduce two new methods on `PeerState` to avoid direct access to
the internal `pending_requests` map.

tnull committed:
Drop `OutboundCRChannel`

The `OutboundChannel` construct simply wrapped `ChannelOrder` which we
can now simply use directly.

tnull committed:
Actually remember the order state in `ChannelOrder`

We here remember and update the order state and channel details in
`ChannelOrder`

tnull committed:
Drop `chain_source` from `LSPS1ServiceHandler`

We previously considered tracking payment confirmations as part of the
handler. However, we can considerably simplify our logic if we stick
with the current approach of having the LSPs track the payment status
and update us when prompted through events.

tnull committed:
Drop `Listen`/`Confirm`/etc from `LiquidityManager`

Now that we don't do on-chain tracking in LSPS1, we can drop quite a few
`LiquidityManager` parameters and generics, which were only added in
anticipation of tracking on-chain state.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
`LSPS1ServiceHandler`: Use `TimeProvider` when creating new orders

Since we by now have the `TimeProvider` trait, we might as well use it
in `LSPS1ServiceHandler` instead of requiring the user to provide a
`created_at` manually.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Require `supported_options` in `LSPS1ServiceConfig`

In the future we might want to inline the fields in `LSPS1ServiceConfig`
(especially once some are added that we'd want to always/never set for
the user), but for now we just make the `supported_options` field in
`LSPS1ServiceConfig` required, avoiding some dangerous `unwrap`s.

tnull committed:
Respond to `GetOrder` requests from our saved state

Previously, we'd use an event to have the user check the order status
and then call back in. As we already track the order status, we here
change that to a model where we respond immediately based on our state
and have the user/LSP update that state whenever it detects a change
(e.g., a received payment, reorg, etc.). In the next commmit we will
add/modify the corresponding API methods to do so.

tnull committed:
Add serialization logic for LSPS1 `PeerState` types

We add the serializations for all types that will be persisted as part
of the `PeerState`.

tnull committed:
Implement `LSPS1ServiceHandler` persistence and state pruning

We follow the model already employed in LSPS2/LSPS5 and implement
state pruning and persistence for `LSPS1ServiceHandler` state.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Drop `lsps1_service` cfg flag

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Add some checks on provided payment details

As per spec, we check that the user provides at least one payment detail
*and* that they don't provide onchain payment details if
`refund_onchain_address` is unset.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Don't hold write lock in `LSPS{1,2}ServiceHandler::peer_disconnected`

.. as there's no need to do so.

tnull committed:
Add `invalid_token_provided` API method

We add a method that allows the LSP to signal to the client the token
they used was invalid.

We use the `102` error code as proposed in
https://github.com/lightning/blips/pull/68.

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Add test case for `invalid_token_provided` flow

We test the just-added API.

Co-authored by Claude AI

tnull committed:
Add test case asserting `LSPS1ServiceState` is persisted across restarts

Co-authored by Claude AI

tnull committed:
Read persisted LSPS1ServiceHandler state on startup

.. we read the persisted state in `LiquidityManager::new`

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Fix clippy lints

tnull committed:
Refactor `ChannelOrder` to use `ChannelOrderState` state machine

This refactors `ChannelOrder` to use an internal state machine enum
`ChannelOrderState` that:
- Encapsulates state-specific data in variants (e.g., `channel_info`
  only available in `CompletedAndChannelOpened`)
- Provides type-safe state transitions
- Replaces the generic `update_order_status` API with specific
  transition methods: `order_payment_received`, `order_channel_opened`,
  and `order_failed_and_refunded`

The state machine has four states:
- `ExpectingPayment`: Initial state, awaiting payment
- `OrderPaid`: Payment received, awaiting channel open
- `CompletedAndChannelOpened`: Terminal state with channel info
- `FailedAndRefunded`: Terminal state for failed/refunded orders

Co-Authored-By: HAL 9000
Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull committed:
Add integration tests for LSPS1 order state transition API

Add two new integration tests to cover the new public API methods:

- `lsps1_order_state_transitions`: Tests the full flow of
  `order_payment_received` followed by `order_channel_opened`,
  verifying that payment states are updated correctly and channel
  info is returned after the channel is opened.

- `lsps1_order_failed_and_refunded`: Tests the `order_failed_and_refunded`
  method, verifying that payment states are set to Refunded.

Co-Authored-By: HAL 9000

tnull committed:
Add integration test for expired order pruning

Add `lsps1_expired_orders_are_pruned_and_not_persisted` test that verifies:

- Orders with expired payment details (expires_at in the past) are
  accessible before persist() is called
- After persist() is called, expired orders in ExpectingPayment state
  are pruned and no longer accessible
- Pruned orders are not recovered after restart, confirming that
  the pruning also removes the persisted state

Co-Authored-By: HAL 9000

tnull committed:
Drop unused `LSPS1OnchainPayment` type

tnull committed:
Add `Hold` payment state per bLIP-51 spec

The bLIP-51 specification defines a `HOLD` intermediate payment state:
- `EXPECT_PAYMENT` -> `HOLD` -> `PAID` (success path)
- `EXPECT_PAYMENT` -> `REFUNDED` (failure before payment)
- `HOLD` -> `REFUNDED` (failure after payment received)

This commit adds the `Hold` variant to `LSPS1PaymentState` and updates
the state machine transitions:

- `payment_received()` now sets payment state to `Hold` (not `Paid`)
- `channel_opened()` transitions payment state from `Hold` to `Paid`
- Tests updated to verify the correct state at each transition

This allows LSPs to properly communicate when a payment has been
received but the channel has not yet been opened (e.g., Lightning
HTLC held, or on-chain tx detected but channel funding not published).

Co-Authored-By: HAL 9000

TheBlueMatt commented on lightning-liquidity/src/lsps1/peer_state.rs:
```diff
@@ -0,0 +1,749 @@
+	) -> ChannelOrder {
+		let state = ChannelOrderState::new(payment_details);
+		let channel_order = ChannelOrder { order_params, state, created_at };
+		self.outbound_channels_by_order_id.insert(order_id, channel_order.clone());
```
might want some kind of size limit here to limit dos, tho we could also just do it at the start of the flow.

tnull replied:
Yeah, I now added a commit following the approach we took for LSPS2, i.e., adding `MAX_TOTAL_PEERS`, `MAX_REQUESTS_PER_PEER` and `MAX_TOTAL_PENDING_REQUESTS` limits.

TheBlueMatt commented on lightning-liquidity/src/lsps1/event.rs:
Probably worth also mentioning that token validation should happen now too.

tnull commented on lightning-liquidity/src/lsps1/peer_state.rs:
```diff
@@ -0,0 +1,749 @@
+		#[cfg(not(feature = "time"))]
+		{
+			// TODO: We need to find a way to check expiry times in no-std builds.
+			all_payment_details_expired = false;
```
I don't think so, as we still want to show the `Failed` state back to the user querying their order status.

tnull commented on lightning-liquidity/src/lsps1/service.rs:
```diff
@@ -313,100 +469,168 @@ where
+	/// Marks an order as paid after payment has been received.
+	///
+	/// This should be called when the LSP detects that a Lightning payment has arrived
+	/// or an on-chain payment has been confirmed.
```
Added.

tnull committed:
Drop unused `LSPS1ServiceEvent::Refund` event

Turns out this was another variant we didn't actually use anywhere. So
we're dropping it.

tnull committed:
Add `onchain_payment_required` method

We previously had no way to reject requests in case the LSP requires
onchain payment while the client not providing
`refund_onchain_address`. Here we add a method allowing to do so.

tnull committed:
Limit pending requests and peers in LSPS1 service

Add per-peer and global rate limiting to `LSPS1ServiceHandler` to
prevent resource exhaustion, mirroring the existing LSPS2 pattern.

Introduce `MAX_PENDING_REQUESTS_PER_PEER` (10),
`MAX_TOTAL_PENDING_REQUESTS` (1000), and `MAX_TOTAL_PEERS` (100000)
constants and enforce them in `handle_create_order_request`. Rejected
requests receive a `CreateOrderError` with
`LSPS0_CLIENT_REJECTED_ERROR_CODE`. A `total_pending_requests` atomic
counter tracks the global count, and a `verify_pending_request_counter`
debug assertion ensures it stays in sync.

Co-Authored-By: HAL 9000

tnull commented on lightning-liquidity/src/lsps1/service.rs:
```diff
@@ -313,100 +469,168 @@ where
+					let msg = LSPS1Message::Response(request_id.clone(), response).into();
+					message_queue_notifier.enqueue(counterparty_node_id, msg);
+					let err = format!("Failed to handle request due to: {}", e);
+					let action = ErrorAction::IgnoreAndLog(Level::Error);
```
Hmm, note that we already silently fail if we can't parse the message at all. This basically follows what we do elsewhere in the codebase. I think we should do https://github.com/lightningdevkit/rust-lightning/issues/3492 as a follow-up to make sure we follow the same approach everywhere. I'll tag that 0.3

tnull committed:
Reject clients if request registration failed (e.g., duplicative Id)

Signed-off-by: Elias Rohrer <dev@tnull.de>

tnull commented on lightning-liquidity/src/lsps1/service.rs:
```diff
@@ -238,20 +317,89 @@ where
 		{
 			let mut outer_state_lock = self.per_peer_state.write().unwrap();
 
+			if !outer_state_lock.contains_key(counterparty_node_id)
```
Done.

TheBlueMatt reviewed (commented):
Still needs the DoS vuln fixed, but mostly looks good.

TheBlueMatt commented on lightning/src/util/ser.rs:
```diff
@@ -1499,6 +1503,39 @@ impl Readable for OutPoint {
+}
+
+impl Readable for Address {
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
```
Should we note `ReadableArgs` this to validate the `Network`?

TheBlueMatt commented on lightning-liquidity/src/lsps1/service.rs:
```diff
@@ -261,50 +409,178 @@ where
-
-				match peer_state_lock.pending_requests.remove(&request_id) {
-					Some(LSPS1Request::CreateOrder(params)) => {
+				let request = peer_state_lock.remove_request(&request_id).map_err(|e| {
```
It does seem kinda weird to me that `remove_request` and `new_order` are separate - to me it seems like one state transition for an order, but we're storing it separately in `PeerState` and removing+re-adding it here.

TheBlueMatt reviewed (commented):
FWIW I'm not really a huge fan of the abstraction between `peer_state` and `service` - the stuff in `peer_state` is use as just a simple type with simple accessors that do relatively little - they'd almost be better off just having everything public so that its clear what's going on from the callsite. Alternatively, moving some of the actual validation into `peer_state` would clean up `service` a lot (making it mostly error conversion and connection-limiting logic, which is fine). Its not really worth changing here, but it does make the code overall feel less readable.

tnull commented on lightning/src/util/ser.rs:
```diff
@@ -1499,6 +1503,39 @@ impl Readable for OutPoint {
+	}
+}
+
+impl Writeable for Bolt11Invoice {
```
Hmm, for the longest time the argument was that we can't implement `Readable`/`Writeable` as the `lightning-invoice->lightning` dependency was inverted. And we also long concluded that it's fine to add serialization for `Offer`,`Refund`, etc. Why can't we just keep it consistent and implement `Readable`/`Writeable` for `Bolt11Invoice` now that we can?

TheBlueMatt reviewed (commented):
This LGTM, feel free to squash.

tnull committed:
Validate all common fields in LSPS1 `is_valid` order check

Add missing cross-validation of `LSPS1OrderParams` against
`LSPS1Options` as required by bLIP-51:

- Check `required_channel_confirmations` >= `min_required_channel_confirmations`
- Check `funding_confirms_within_blocks` >= `min_funding_confirms_within_blocks`
- Check total channel balance (`lsp_balance_sat` + `client_balance_sat`)
  is within [`min_channel_balance_sat`, `max_channel_balance_sat`],
  using `checked_add` to guard against overflow

Co-Authored-By: HAL 9000

tnull committed:
Reset `persistence_in_flight` counter on error in LSPS1/LSPS2

Previously, if any `.await?` in the persist loop returned an error, the
`?` would propagate out of `persist()` before reaching the `fetch_sub`
at the end of the loop. This left the counter permanently > 0, causing
all subsequent `persist()` calls to early-return and effectively
disabling persistence for the lifetime of the handler.

Fix this by extracting the loop into `do_persist()` and unconditionally
resetting the counter via `store(0, Release)` in the outer `persist()`
after `do_persist()` returns, regardless of success or failure.

Co-Authored-By: HAL 9000

TheBlueMatt commented:
`check_commits` is failing.

tnull commented on lightning-liquidity/src/lsps1/peer_state.rs:
```diff
@@ -0,0 +1,781 @@
+			})
+			.count();
+		pending_requests + unpaid_orders
+	}
```
Known, wontfix, as mentioned above.

tnull commented on lightning-liquidity/src/lsps1/service.rs:
```diff
@@ -178,11 +122,160 @@ where
+		let res = self.do_persist().await;
+		debug_assert!(res.is_err() || self.persistence_in_flight.load(Ordering::Acquire) == 0);
+		self.persistence_in_flight.store(0, Ordering::Release);
+		res
```
I think this okay for now, our async persistence has a bunch of these race-y edge cases anyways.

tnull commented:
We might want to backport 47e5c04f64492ade647652463dd6e3435f1aa815 to 0.2

TheBlueMatt commented:
Backported in #4706
